### PR TITLE
Dual arms with MoveIt

### PR DESCRIPTION
**Pull request by AndyZe**
### Description

A tutorial for dual-arm setup, planning, and execution with MoveIt.

### Testing

`ros2 launch dual_arm_panda_moveit_config demo.launch.py`

@jackcenter @corycrean @[pabloinigoblasco](https://github.com/pabloinigoblasco)

### CODE REVIEW

File: doc/examples/dual_arms/dual_arm_panda_moveit_config/config/moveit_controllers.yaml [hunk 1]
```diff
@@ -4,34 +4,4 @@ trajectory_execution:
...
-  default: true
-  joints:
-    - right_panda_joint1
-    - right_panda_joint2
-    - right_panda_joint3
-    - right_panda_joint4
-    - right_panda_joint5
-    - right_panda_joint6
-    - right_panda_joint7
+moveit_controller_manager: moveit_ros_control_interface/MoveItMultiControllerManager
```
Comment (JafarAbdi) [inline doc/examples/dual_arms/dual_arm_panda_moveit_config/config/moveit_controllers.yaml:7]: I accidentally used `MoveItMultiControllerManager`, `moveit_ros_control_interface/MoveItControllerManager` should work too

**Comment by AndyZe** [comment]
This tutorial is functional now (thanks @JafarAbdi !) Some things it still needs to get merged:

- tutorial text
- clean up of extraneous files in the `dual_arm_panda_moveit_config` folder. I think there are some xacro and yaml files that aren't used. Let's strip this down to the bare minimum

**Comment by pabloinigoblasco** [comment]
Hello guys.

Is this something that should be already working or it is "work in progress"?

I managed to build everything in rolling-focal but in runtime there were problems loading controllers.
I assume you are working on jammy for your local environment (at least CI is based on jammy).

If this is not yet ready, is there anything we could do to help?

Regards.

**Comment by AndyZe** [comment]
Hi Pablo, it is working and ready to merge as far as I know. You're correct that is on Jammy though.

I know the same concepts will work on Galactic, however, because I copied it from a Galactic project.

**Comment by pabloinigoblasco** [comment]
> Hi Pablo, it is working and ready to merge as far as I know. You're correct that is on Jammy though.
> 
> I know the same concepts will work on Galactic, however, because I copied it from a Galactic project.

Hello @AndyZe , we found some problems to build the demo for galactic, but disabling a few parts of the code and other examples we managed to build it.

There is some important missing file for the dual arm demo "config/cartesian_limits.yaml" that breaks the execution of the demo.launch.py, but I picked up some of the existing ones from other examples and it worked.

We are also getting problems with the moveit rviz displays, we get "NO PLANNING LIBRARY LOADED", so we cannot test moveit from rviz directly.

At the current moment we are trying to use both arms from code. We will make you know how it goes!
Thanks.

**Comment by AndyZe** [comment]
> There is some important missing file for the dual arm demo "config/cartesian_limits.yaml" that breaks the execution of the demo.launch.py...

This is probably because you're trying to load the Pilz motion planner in your launch files. I'd suggest to search for and delete "pilz".

**Comment by pabloinigoblasco** [comment]
> > There is some important missing file for the dual arm demo "config/cartesian_limits.yaml" that breaks the execution of the demo.launch.py...
> 
> This is probably because you're trying to load the Pilz motion planner in your launch files. I'd suggest to search for and delete "pilz".

Thanks Andy. I will have a look to that.

Also, making you know that we could already move both arms vía moveit2_cpp and it is working properly. Thanks.
We can work on top of this for our examples!

**Comment by AndyZe** [comment]
Alright. The orocos_kdl_vendor issue popped up during the transition from Ubuntu 20 -> Ubuntu 22. I'm pretty sure it would go away if you could get on Ubuntu 22 and Humble or Rolling (or stay on Ubuntu 20 and Galactic). Sorry I don't have better suggestions. Feel free to create MoveIt issues for these problems because I don't think they're related to this PR.

**Comment by AndyZe** [comment]
Have you tried making one joint group for both arms? Something like this in `panda.srdf`:

  <group name="left_panda_arm">
    <chain base_link="left_panda_link0" tip_link="left_panda_link8"/>
  </group>
  <group name="right_panda_arm">
    <chain base_link="right_panda_link0" tip_link="right_panda_link8"/>
  </group>
  <group name="both_arms">
    <group name="left_panda_arm">
    <group name="right_panda_arm">
  </group>
  
  I found this on answers.ros.org:  https://answers.ros.org/question/370136/planning-for-a-dual-arm-robot-in-moveit/

**Comment by henrygerardmoore** [comment]
When going through the example I run into
```
ws_moveit2$ ros2 launch dual_arm_panda_moveit_config demo.launch.py
[INFO] [launch]: All log files can be found below /home/henry/.ros/log/2022-06-08-09-23-05-301156-henry-picknik-laptop-118051
[INFO] [launch]: Default logging verbosity is set to INFO
WARNING:root:Cannot infer URDF from `/home/henry/ws_moveit2/install/moveit2_tutorials/share/dual_arm_panda_moveit_config`. -- using config/dual_arm_panda.urdf
WARNING:root:Cannot infer SRDF from `/home/henry/ws_moveit2/install/moveit2_tutorials/share/dual_arm_panda_moveit_config`. -- using config/dual_arm_panda.srdf
[ERROR] [launch]: Caught exception in launch (see debug for traceback): Caught exception when trying to load file of format [py]: "File /home/henry/ws_moveit2/install/moveit2_tutorials/share/dual_arm_panda_moveit_config/config/cartesian_limits.yaml doesn't exist"
```
Did I miss something?

**Comment by AndyZe** [comment]
That's the same error Pablo had. OK, I must be using an older version of MoveIt (specifically moveit_configs_utils). OK, thanks. I'll try to fix this.

**Review by henrygerardmoore — APPROVED** [review]
Tutorial seems good to go other than removing the hand planning groups!

**Comment by pabloinigoblasco** [comment]
> Have you tried making one joint group for both arms? Something like this in `panda.srdf`:
> 
> ```
>   <group name="left_panda_arm">
>     <chain base_link="left_panda_link0" tip_link="left_panda_link8"/>
>   </group>
>   <group name="right_panda_arm">
>     <chain base_link="right_panda_link0" tip_link="right_panda_link8"/>
>   </group>
>   <group name="both_arms">
>     <group name="left_panda_arm">
>     <group name="right_panda_arm">
>   </group>
> ```
> 
> I found this on answers.ros.org: https://answers.ros.org/question/370136/planning-for-a-dual-arm-robot-in-moveit/

I am trying this. I can move both arms using the setJointValueTarget in one single execution request.
I see that this approach forces you to have synchronized motions between arms (same duration of motion of both arms). And this would not be generic enough for totally independent tasks for each arm.

I tried this. I am getting the following results:

```
**both_group**.setJointValueTarget(..) -> **success**, both arms move at the same time
left_group.setTargetPose(A, ee_left) -> success, left arm moves
right_group.setTargetPose(B, ee_right) -> success, right arm moves

**both_group**.setTargetPose(A, ee_left)-> plan failure
**both_group**.setTargetPose(A, ee_left); both_group.setTargetPose(A, ee_right) -> plan failure
```

By the moment for end effector dual motion planning it is not working for me.
I have to think about this, but I do not see the logic right now.

**Comment by pabloinigoblasco** [comment]
In this thread we have been mostly discussing about the dual-arm single-robot (one move-group  node one controller-manager node) solution.

- My current view is that this solution is not enough to execute different unsynchronized tasks with each arm because one single move_groups looks like are not able to handle in concurrently multiple requests to different groups.

I think the most natural solution would be a **multiple-robot solution**. We attempted that a couple of times in the last few months but we did not success.  I think this is a solution required but not still available/tested. 


**There are like a few possible strategies:**
   - using namespaces for ros2_control/move_group/move_group_interface client (two move-groups nodes, two controller managers nodes in namespaces) - Didn't work for us 
   - putting everything in the same root namespace using topic/action remapping (one single controller manager two-move groups nodes) - Didn't work for us

**We got general conclusions :**
- **ros2_control likes topic/action remapping for multiple arms.** Namespaces functionality for controllers does not look stable enough at the current moment (at least in galactic) . We tried this, but it didnt work. Nonetheless, we do not think this is a good solution (one single central controller manager handling two different robots)  

- **moveit likes/need namespaces for multiple move_groups.**  multiple move_group nodes do not work well with topic/actions remapping in the same root namespace, there are collisions on topics and actions you cant remap.

File: doc/examples/dual_arms/dual_arms_tutorial.rst [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+- Enumerate the controllers we will need for both arms in the ros2_control config file, ``ros2_controllers.yaml``. Make sure these controllers are launched from ``demo.launch.py``.
+
+- Define joint groups for each arm in ``panda.srdf``. This tells MoveIt which joints constitute each arm. The joint groups are named ``left_panda_arm`` and ``right_panda_arm``. Also define an end effector for each arm.
+
+- Define a kinematics solver for each arm in ``kinematics.yaml``.
+
+- Define the controllers which MoveIt can execute trajectories with in ``moveit_controllers.yaml``. Here we have a trajectory controller for each arm.
+
+- Also in ``moveit_controllers.yaml``, define the controller management strategy MoveIt will use. The simplest option from a configuration standpoint is either ``moveit_ros_control_interface/MoveItMultiControllerManager`` or ``moveit_ros_control_interface/MoveItControllerManager``. You can also use a ``SimpleControllerManager`` although it requires additional namespacing and additional enumeration of the joints.
```
Comment (JafarAbdi) [inline doc/examples/dual_arms/dual_arms_tutorial.rst:41]:
```suggestion
- Also in ``moveit_controllers.yaml``, define the controller management strategy MoveIt will use. The simplest option from a configuration standpoint is either ``moveit_ros_control_interface/MoveItMultiControllerManager`` or ``moveit_ros_control_interface/MoveItControllerManager``. You can also use a ``moveit_simple_controller_manager/MoveItSimpleControllerManager`` although it requires additional namespacing and additional enumeration of the joints.
```

**Review by JafarAbdi — APPROVED** [review]

**Pull request merged**